draft-barnes-acme.md

```diff
@@ -218,304 +219,275 @@ In this section, we describe the four certificate management functions that ACME
+## Errors
 
-| Request              | Response      |
-|:=====================|:==============|
-| challengeRequest     | challenge     |
-| authorizationRequest | authorization |
-| certificateRequest   | certificate   |
-| revocationRequest    | revocation    |
-| statusRequest        | (any)         |
+Errors can be reported in ACME both at the HTTP layer and within ACME payloads.  ACME servers can return responses with an HTTP error response codes (4XX or 5XX).  For example:  If the client submits a request using a method not allowed in this document, then the server MAY return status code 405 (Method Not Allowed).
 
-## Signatures
+When the server responds with an error status, it SHOULD provide additional information using problem document {{I-D.ietf-appsawg-http-problem}}.  The "type", "detail", and "instance" fields MUST be populated.  To facilitate automatic response to errors, this document defines the following standard tokens for use in the "type" field (within the "urn:acme:" namespace):
 
-ACME uses a simple JSON-based structure for encoding signatures, based on the JSON Web Signature structure.  An ACME signature is a JSON object, with the following fields:
+| Code            | Semantic                                                 |
+|:================|:=========================================================|
+| malformed       | The request message was malformed                        |
+| unauthorized    | The client lacks sufficient authorization                |
+| serverInternal  | The server experienced an internal error                 |
+| badCSR          | The CSR is unacceptable (e.g., due to a short key)       |
 
-alg (required, string):
-: A token indicating the cryptographic algorithm used to compute the signature {{I-D.ietf-jose-json-web-algorithms}}. (MAC algorithms such as "HS*" MUST NOT be used.)
+Authorization and challenge objects can also contain error information to indicate why the server was unable to validate authorization.
 
-sig (required, string):
-: The signature, base64-encoded.
+TODO: Flesh out errors and syntax for them
 
-nonce (required, string):
-: A signer-provided random nonce of at least 16 bytes, base64-encoded.  (For anti-replay.)
+## Key Authorization
 
-jwk (required, object):
-: A JSON Web Key object describing the key used to verify the signature {{I-D.ietf-jose-json-web-key}}.
+The key authorization process establishes the authorization of an account key pair to manage certificates for a given identifier.  This process must assure the server of two things: First, that the client controls the private key of the key pair, and second, that the client holds the identifier in question.  This process may be repeated to associate multiple identifiers to a key pair (e.g., to request certificates with multiple identifiers), or to associate multiple key pairs with an identifier (e.g., for load balancing).
```

Comment:
I assume you're talking about authorizing N different account keys for a given identifier? Yes, I think that's valuable and will be pretty crucial for advanced users (i.e. hosting providers).

I think to support that we will probably want to flesh out the notion of an 'account', and say that one account can have multiple account keys, and can also have multiple identifiers for which it is authorized. Otherwise we wind up with a messy unmanageable model of N * M authorizations for an organization with N staff and M identifiers under their control.